homeassistant/components/sensor/ambient_station.py

```diff
@@ -133,80 +137,46 @@ def unit_of_measurement(self):
+        self._data.update()
+        latest_data = self._data.data[self._station.mac_address]
 
-        _LOGGER.debug("Sensor: %s | Data: %s", self._name, self._state)
+        try:
+            self._state = latest_data[self._condition]
+        except KeyError:
+            _LOGGER.warning('No data for condition: %s', self._condition)
 
 
 class AmbientStationData:
-    """Class to interface with ambient-api library."""
-
-    def __init__(self, hass, api_key, app_key):
-        """Initialize station data object."""
-        self.hass = hass
-        self._api_keys = {
-            'AMBIENT_ENDPOINT':
-            'https://api.ambientweather.net/v1',
-            'AMBIENT_API_KEY': api_key,
-            'AMBIENT_APPLICATION_KEY': app_key,
-            'log_level': 'DEBUG'
-        }
-
-        self.data = None
-        self._station = None
-        self._api = None
-        self._devices = None
-        self.connect_success = False
-
-        self.get_data = Throttle(SCAN_INTERVAL)(self.async_update)
-        self._connect_api()     # attempt to connect to API
-
-    async def async_update(self):
-        """Get new data."""
-        # refresh API connection since servers turn over nightly
-        _LOGGER.debug("Getting new data from server")
-        new_data = None
-        await self.hass.async_add_executor_job(self._connect_api)
-        await asyncio.sleep(2)   # need minimum 2 seconds between API calls
-        if self._station is not None:
-            data = await self.hass.async_add_executor_job(
-                self._station.get_data)
-            if data is not None:
-                new_data = data[0]
-                self.data = new_data
-            else:
-                _LOGGER.debug("data is None type")
-        else:
-            _LOGGER.debug("Station is None type")
-
-        return new_data
-
-    def _connect_api(self):
-        """Connect to the API and capture new data."""
-        from ambient_api.ambientapi import AmbientAPI
-
-        self._api = AmbientAPI(**self._api_keys)
-        self._devices = self._api.get_devices()
-
-        if self._devices:
-            self._station = self._devices[0]
-            if self._station is not None:
-                self.connect_success = True
-        else:
-            _LOGGER.debug("No station devices available")
+    """Define an object to retrieve data from the Ambient API."""
+
+    def __init__(self, api):
+        """Initialize the station data object."""
+        self._api = api
+        self.data = {}
+        self.stations = []
+
+    @Throttle(DEFAULT_SCAN_INTERVAL)
+    def update(self):
+        """Get new data for all stations."""
+        # Ambient's API has really aggressive rate limiting (no more than 1
+        # request per second), so throughout this method, we make sure to sleep
+        # at least 1 second before firing further requests:
+        self.stations = self._api.get_devices()
+        sleep(1)
```

Comment:
We're not allowed to sleep in our thread pool. That was the whole purpose of the asyncio use that is removed here. Was there anything else that you considered incorrect about it?

Reply:
Ah, didn't know that; thank you for helping me! I'll add it back in.

Reply:
Check 2e6651659b89 for me?

FYI, once Ambient fixes their socket-based API, the long-term goal is to make the underlying library responsible for all the sleeping. Round 2. 😄

Reply:
Looks good!